src/serprog.c

```diff
@@ -0,0 +1,510 @@
+    unsigned char params[6];
+    write_le24(params, len);
+    write_le24(params + 3, len);
+    if (perform_serprog_cmd_full(pgm, S_CMD_O_SPIOP, params, sizeof(params), tx, len, rx, len) != 0) {
+        return -1;
+    }
+
+    return len;
+}
+
+static bool is_serprog_cmd_supported(const unsigned char *cmd_bitmap, unsigned char cmd) {
+    return (cmd_bitmap[cmd / 8] >> (cmd % 8)) & 1;
+}
+
+// programmer lifecycle handlers
+
+static int serprog_open(PROGRAMMER *pgm, const char *pt) {
+    const char *port_error =
+        "unknown port specification, "
+        "please use the format /dev/ttyACM0\n";
+    char port_default[] = "/dev/ttyACM0";
+    char *serialdev;
+    char *port = mmt_strdup(pt);
+
+    if (str_eq(port, "unknown")) {
+        port = port_default;
+    }
+
+    serialdev = strtok(port, ",");
+    if (!serialdev) {
+        pmsg_error("%s", port_error);
+        return -1;
+    }
+
+    union pinfo pinfo;
+    pgm->port = port;
+    pinfo.serialinfo.baud = pgm->baudrate? pgm->baudrate: 115200;
+    pinfo.serialinfo.cflags = SERIAL_8N1;
+    if (serial_open(port, pinfo, &pgm->fd)==-1) {
+        return -1;
+    }
+
+    unsigned char buf[32];
+
+    // sync
+    memset(buf, 0, sizeof(buf));
+    if (perform_serprog_cmd(pgm, S_CMD_SYNCNOP, NULL, 0, buf, 1) != 1 || buf[0] != S_ACK) {
+        pmsg_error("cannot sync; is this a serprog programmer?\n");
+        return -1;
+    }
+
+    // get command bitmap
+    memset(my.cmd_bitmap, 0, sizeof(my.cmd_bitmap));
+    if (perform_serprog_cmd(pgm, S_CMD_Q_CMDMAP, NULL, 0, my.cmd_bitmap, 32) != 0) {
+        pmsg_error("cannot get list of supported serprog commands\n");
+        return -1;
+    }
+
+    // get protocol version
+    memset(buf, 0, sizeof(buf));
+    if (!is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_Q_IFACE)
+        || perform_serprog_cmd(pgm, S_CMD_Q_IFACE, NULL, 0, buf, 2) != 0) {
+        pmsg_error("cannot get serprog protocol version\n");
+        return -1;
+    }
+    if (read_le16(buf) != 0x01) {
+        pmsg_error("unsupported serprog protocol version: %d\n", read_le16(buf));
+        return -1;
+    }
+
+    pmsg_info("serprog protocol version: %d\n", read_le16(buf));
+
+    // get programmer name
+    if (is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_Q_PGMNAME)) {
+        memset(buf, 0, sizeof(buf));
+        if (perform_serprog_cmd(pgm, S_CMD_Q_PGMNAME, NULL, 0, buf, 16) != 0) {
+            pmsg_error("cannot get programmer name\n");
+            return -1;
+        }
+        pmsg_info("programmer name: %s\n", buf);
+    }
+
+    // check if required commands are supported
+    if (!is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_O_SPIOP)) {
+        pmsg_error("the %s programmer does not support SPI operations\n", pgmid);
+        return -1;
+    }
+
+    if (!is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_S_CS_MODE)) {
+        pmsg_error("the %s programmer does not support setting the CS mode\n", pgmid);
+        return -1;
+    }
```

Comment:
Is it necessary to balk if the programmer cannot set CS? Maybe the user doesn't need to set it?

Reply:
This line does not check if it is possible to change the CS to a different pinout. This is done only if the user requests another CS than CS_0. This line instead checks if the 0x18 Set CS Mode command is supported. This command is required as it allows to keep the chip select high for the whole programming operation and not only for each SPI message as in the default setting.